### PR TITLE
feat(network): integrate Astar network

### LINKED ISSUE #6148: [Bounty] Integrate Astar, a Multi-Chain Smart Contract into Mask Plugin
## Catalogue

- Platform
  - [ ] Gitcoin
  - [ ] Hackerlink
  - [x] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] $2000 USD or $MASK equivalent 

## Part 1: Problem Description

Astar Network connects the Polkadot ecosystem to Ethereum, Cosmos, and all major Layer 1 Blockchains. Astar Network supports dApps using multiple virtual machines WASM and EVM and offers the best technology and financial incentives via Build2Earn and Astar Incubation Program.

**Solution** 
Integrate Astar into Mask Plugin

**To Do** 
- Migrate the following applications to ASTAR：redpocket, balancerchecker, multical, tips.
- Basis function: connect, change wallet, presentation of visual elements, toast

**RPC**
- Network Name
- Astar Network
- Network URL
- https://evm.astar.network/
- Chain ID
- 592

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- [UI](https://astar.network/)
- https://astar.network/
- https://github.com/AstarNetwork

## Part 3: Contact and Support
@yisiliu @tzhan28 @LindaMask for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q,
Astar Discord: https://discord.com/invite/Z3nC9U4

DM on telegram @lin_dao and she will create a dedicated group with the Mask team to start the bounty process

Submission Requirements
Please include demo to show the integration is working.
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by developerfred**
## Description

Closes  #6148

### DEX 

 -  Arthswap ✅ 

### contracts 

 - multicall ✅ 
 - balanceChecker ✅ 

### Token List
- [x] https://github.com/DimensionDev/Mask-Token-Lists/issues/35

### screen test
https://youtu.be/ZkTulTeVPwI

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by guanbinrui** [comment]
Hi @DimensionDev/dev-smartcontract, please deploy mask contracts on Astar Network.

**Review by liqingnz — DISMISSED** [review]

**Comment by Jack-Works** [comment]
there are many merge conflicts 👀

File: packages/web3-shared/evm/pipes/index.ts [hunk 1]
```diff
@@ -153,6 +163,10 @@ export const resolveChainColor = createLookupTableResolver<ChainId, string>(
         [ChainId.Harmony_Test]: 'rgb(48, 153, 242)',
         [ChainId.Aurora_Testnet]: 'rgb(112, 212, 74)',
         [ChainId.Astar]: 'rgb(36, 150, 238)',
+        [ChainId.Fuse]: 'rgb(255 255 255)',
+        [ChainId.Boba]: 'rgb(255 255 255)',
+        [ChainId.Metis]: 'rgb(255 255 255)',
+        [ChainId.Optimistic]: 'rgb(255 255 255)',
```
Comment (septs) [inline packages/web3-shared/evm/pipes/index.ts]:
```suggestion
        [ChainId.Fuse]: 'rgb(255, 255, 255)',
        [ChainId.Boba]: 'rgb(255, 255, 255)',
        [ChainId.Metis]: 'rgb(255, 255, 255)',
        [ChainId.Optimistic]: 'rgb(255, 255, 255)',
```

**Review by liqingnz — DISMISSED** [review]

**Review by liqingnz — DISMISSED** [review]

**Review by liqingnz — DISMISSED** [review]

File: packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx [hunk 2]
```diff
@@ -23,12 +23,13 @@ import { WalletRPC } from '../../../Wallet/messages'
...
 import type { GasSettingProps } from './types'
 import { useGasSettingStyles } from './useGasSettingStyles'
-import { pow10 } from '@masknet/web3-shared-base'
+import { pow10, scale10 } from '@masknet/web3-shared-base'
 
 const minGasPriceOfChain: ChainIdOptionalRecord<BigNumber.Value> = {
     [ChainId.BSC]: pow10(9).multipliedBy(5), // 5 Gwei
     [ChainId.Conflux]: pow10(9).multipliedBy(5), // 5 Gwei
     [ChainId.Matic]: pow10(9).multipliedBy(30), // 30 Gwei
+    [ChainId.Astar]: scale10(9).multipliedBy(5), // 5 Gwei
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx]:
```suggestion
    [ChainId.Astar]: scale10(5, 9), // 5 Gwei
```

Comment (guanbinrui) [reply packages/mask/src/plugins/Wallet/SNSAdaptor/GasSettingDialog/Prior1559GasSetting.tsx]: The `scale10` has got a bug in it. Fixed by #6334.

**Review by guanbinrui — DISMISSED** [review]

**Review by guanbinrui — DISMISSED** [review]

**Comment by AinuX** [comment]
@developerfred 

1. Unable to load token list and trans history
image

2. Crash when click "Swap" in application board.

https://user-images.githubusercontent.com/2582974/170695006-2225589b-c69b-4e44-b604-2dba43f7fdf6.mp4

> Failed to create token.

Error stack:

<pre>Error: Failed to create token.
    at createNativeToken (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/414.js:37457:38)
    at chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/chunk.6056.js:4202:195
    at Oj (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:248:332)
    at Gk (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:291:111)
    at Vk (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:288:358)
    at Ok (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:286:398)
    at Dk (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:278:439)
    at ig (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:133:105)
    at Vk (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:288:419)
    at Ok (chrome-extension://cabdakbbdapmjcmcdajnlilnpofjmddb/js/npm.react-dom.js:286:398)</pre>

3. LuckyDrop is not work 

https://user-images.githubusercontent.com/2582974/170695150-c37a821e-9efb-4347-8a9f-c5d12d10d55a.mp4

**Comment by AinuX** [comment]
@developerfred 

https://user-images.githubusercontent.com/2582974/175335824-da54fbc2-5afb-4841-b555-1abafef9e974.mp4

1. Chain icon is incorrect
2. Native token balance and symbol is incorrect. 

image
3. Balance in dashboard is incorrect

image
4.  Native token do not need to import, and token icon is missing in token list

5.  Still no trans history

https://user-images.githubusercontent.com/2582974/175336613-248014fe-a0d7-4c3a-bdd0-582d7c1399c5.mp4

6.  Token transfer not work 
7.  Gas setting not work

https://user-images.githubusercontent.com/2582974/175337227-d41b7e81-c9ef-4dde-b193-f398b66b3788.mp4

8.  LuckyDrop not work

**Comment by developerfred** [comment]
@AinuX  3 waiting deploy balanceChecker contract ref https://github.com/DimensionDev/Maskbook/issues/6645

**Comment by developerfred** [comment]
@AinuX

**QA Round 1:**
- [x] 1 - Unable to load token list and trans history **[Fixed]**   ✅ 
- [x] 2 - Crash when click "Swap" in application board.  **[Fixed]**   ✅ 
- [x] LuckyDrop is not work **[Fixed]**   ✅ 


**QA Round 2:**

- [x] 1 -  Chain icon is incorrect **[Fixed]**   ✅ 
- [x] 2 -  Native token balance and symbol is incorrect. **[Fixed]**   ✅ 
- [x] 3. Balance in dashboard is incorrect **[Fixed]**   ✅ 
- [x] 4. Native token do not need to import, and token icon is missing in token list and  **[Fixed]**   ✅ 
- [x] 5. Still no trans history **[Fixed]**   ✅ 
- [x] 6. Token transfer not work **[Fixed]**   ✅ 
- [x] 7. Gas setting not work **[Fixed]**   ✅ 
- [x] 8. LuckyDrop is not work  **[Fixed]**   ✅ 

 
**Shoots** 

Screenshot 2022-06-29 at 21 06 51



https://user-images.githubusercontent.com/57037080/176541620-36c11cd4-b882-4a23-9c42-36d24588a167.mov


Screenshot 2022-06-29 at 21 28 47






Screenshot 2022-06-29 at 21 54 40

**Comment by AinuX** [comment]
@developerfred 

Still can't get token balance.

https://user-images.githubusercontent.com/2582974/177265250-f0883910-30d0-44b3-9584-b88caea44933.mp4

**Comment by developerfred** [comment]
> @developerfred
> 
> Still can't get token balance.
> 
>  iShot_2022-07-05_14.37.50.mp4

Screenshot 2022-07-05 at 23 08 56

File: packages/web3-constants/evm/rpc.json [hunk 3]
```diff
@@ -91,6 +92,7 @@
         "Aurora_Testnet": [0, 0, 0, 0, 0],
         "Conflux": [0, 0, 0, 0, 0],
         "Harmony": [0, 0, 0, 0, 0],
-        "Harmony_Test": [0, 0, 0, 0, 0]
+        "Harmony_Test": [0, 0, 0, 0, 0],
+        "Astar": [0, 1, 2, 3, 4]
```
Comment (guanbinrui) [inline packages/web3-constants/evm/rpc.json]:
```suggestion
        "Astar": [0, 0, 0, 0, 0]
```

**Comment by guanbinrui** [comment]
Hi guys, @developerfred @AinuX. I have added Astar RPC into CI envs.

File: packages/plugins/EVM/src/state/Hub/hub.ts [hunk 4]
```diff
@@ -118,7 +119,8 @@ class Hub implements EVM_Hub {
         })
         try {
             const isEIP1559 = chainResolver.isSupport(options.chainId, 'EIP1559')
-            if (isEIP1559) return await MetaSwap.getGasOptions(options.chainId)
+            if (isEIP1559 && chainId !== 592) return await MetaSwap.getGasOptions(options.chainId)
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Hub/hub.ts:122]: Add features in `chains.json` for Astar.

File: packages/plugins/EVM/src/state/Hub/hub.ts [hunk 5]
```diff
@@ -118,7 +119,8 @@ class Hub implements EVM_Hub {
...
+            if (chainId === 592) return await AstarGas.getGasOptions(options.chainId)
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Hub/hub.ts]: Move this first place. Meanwhile, please avoid using magic code.

File: packages/web3-providers/src/astar/index.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                estimatedBaseFee: fromWei(result.estimatedBaseFee) ?? '0',
+                estimatedSeconds: result.timestamp ?? 0,
+                suggestedMaxFeePerGas: tip.slow ?? '0',
+                suggestedMaxPriorityFeePerGas: fromWei(priorityFeePerGas.slow) ?? '0',
+            },
+        }
+    }
+}
+
+export function fromWei(value: string): string {
```
Comment (guanbinrui) [inline packages/web3-providers/src/astar/index.ts]: Please use `formatWeiToGwei`.

File: packages/web3-constants/evm/ito.json [hunk 7]
```diff
@@ -158,6 +164,7 @@
         "Aurora_Testnet": "",
         "Conflux": "",
         "Harmony": "https://api.thegraph.com/subgraphs/name/dimensiondev/mask-ito-harmony",
-        "Harmony_Test": ""
+        "Harmony_Test": "",
+        "Astar": "https://api.thegraph.com/subgraphs/name/dimensiondev/mask-ito-astar"
```
Comment (UncleBill) [inline packages/web3-constants/evm/ito.json]: We don't deployed subgraph for Astar yet, nor Harmony.

image

But I think you can keep this config.

cc @guanbinrui

File: packages/plugins/EVM/src/state/Hub/hub.ts (hunk 5 above)
Comment (developerfred) [reply packages/plugins/EVM/src/state/Hub/hub.ts]: Ok

**Comment by AinuX** [comment]
@developerfred 

1. Balance in token list not work 
2. Swap not work 
3. Decimal issue when select a Non-Native token

https://user-images.githubusercontent.com/2582974/179884216-d7d76ce9-7933-483f-92d9-8cd48aeecb67.mp4

4.  Fail to transfer token
image

File: .gitpod.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# This configuration file was automatically generated by Gitpod.
+# Please adjust to your needs (see https://www.gitpod.io/docs/config-gitpod-file)
+# and commit this file to your remote git repository to share the goodness with others.
+
+tasks:
```
Comment (Jack-Works) [inline .gitpod.yml]: I wonder do we really need this? You cannot debug from the cloud because this is a browser extension.

Comment (developerfred) [reply .gitpod.yml]: we not need this, I will remove.

File: packages/mask/src/plugins/Wallet/services/send.ts [hunk 9]
```diff
@@ -202,6 +202,22 @@ export async function sendPayload(payload: JsonRpcPayload, options?: Options) {
                 if (options?.popupsWindow) openPopupWindow()
                 return
             }
+
+            if (options?.chainId === 592) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/services/send.ts]:
```suggestion
            if (options?.chainId === ChainId.Astar) {
```

File: packages/plugins/EVM/src/state/Hub/hub.ts [hunk 10]
```diff
@@ -91,7 +92,8 @@ class Hub implements EVM_Hub {
...
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Hub/hub.ts]:
```suggestion
            if (isEIP1559 && chainId !== ChainId.Astar) return await MetaSwap.getGasOptions(options.chainId)
            if (chainId === ChainId.Astar) return await AstarGas.getGasOptions(options.chainId)
```

**Review by guanbinrui — DISMISSED** [review]

**Review by guanbinrui — APPROVED** [review]

**Comment by developerfred** [comment]
Uhulll @guanbinrui @nuanyang233

**Pull request merged**